File: Program's_Contributed_By_Contributors/C++/mergesort.cpp
#include <iostream>
using namespace std;

void Merge(int *arr, int s, int e){

    int mid = s + (e-s)/2;

    int len1 = mid - s + 1;
    int len2 = e - mid;

    int *first = new int[len1];
    int *second = new int[len2];

    int mainArrayIndex = s;
    for(int i=0;i<len1;i++){
        first[i] = arr[mainArrayIndex++];
    }

    mainArrayIndex = mid+1;
    for(int i=0;i<len2;i++){
        second[i] = arr[mainArrayIndex++];
    }

    int index1 = 0;
    int index2 = 0;
    mainArrayIndex = s;

    while(index1 < len1 && index2 < len2){
        if(first[index1] < second[index2])
            arr[mainArrayIndex++] = first[index1++];
        else
            arr[mainArrayIndex++] = second[index2++];
    }

    while(index1 < len1)
        arr[mainArrayIndex++] = first[index1++];

    while(index2 < len2)
        arr[mainArrayIndex++] = second[index2++];
}

void MergeSort(int *arr, int s, int e){

    if(s>=e)
        return;

    int mid = (s+e)/2;

    //Left wala part
    MergeSort(arr, s, mid);

    //Right wala part
    MergeSort(arr, mid+1, e);

    //Merging the sorted arrays
    Merge(arr,s,e);
}

int main(){
    int arr[5] = {13,56,2,4,15};
    int n =  5;

    MergeSort(arr,0,n-1);

    for(int i=0;i<n;i++)
        cout << arr[i] << " ";

    cout << endl;

    return 0;
}

File: Program's_Contributed_By_Contributors/C++/Merge_Sort.java
import java.lang.*;

class Algo{

    public void mergeSort(int arr[], int start, int end)
    {
        if(end <= start)
            return;
        
        int mid = (start + (end-start)/2);
        mergeSort(arr, start, mid);
        mergeSort(arr, mid+1, end);
        merge(arr, start, mid, end);
    }

    public void merge(int arr[], int start, int mid, int end)
    {
        int[] mergedArray = new int[end-start+1];

        int i = start;
        int j = mid + 1;
        int k = 0;

        while(i<=mid && j<=end)
        {
            if(arr[i] < arr[j])
                mergedArray[k++] = arr[i++];
            else
                mergedArray[k++] = arr[j++];
        }

        while(i<=mid)
            mergedArray[k++] = arr[i++];
        

        while(j<=end)
            mergedArray[k++] = arr[j++];

        for(int p=0,q=start;p<mergedArray.length;p++,q++)
        {
            arr[q] = mergedArray[p];
        }
        
    }
}

public class Merge_Sort{
    public static void main(String args[])
    {
        int n[] = {9,8,7,6,5,4,3,2,1};

        System.out.print("Unsorted Array: ");
        for(int i=0;i<n.length;i++)
        {
            System.out.print(n[i]+" ");
        }

        Algo a = new Algo();

        a.mergeSort(n, 0, n.length-1);

        System.out.print("\nSorted Array: ");
        for(int i=0;i<n.length;i++)
        {
            System.out.print(n[i]+" ");
        }
    }
}